language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -82,79 +115,154 @@ mergePackageMap ps = foldM merge Map.empty ps
-            | modRef@(pkgRef, modName) <- Set.toList ((PRSelf, curModName) `Set.delete` Set.unions refs)
-            , let pkgRefStr = case pkgRef of
-                    PRSelf -> ""
-                    PRImport pkgId ->
-                      -- If the imported package has a symbolic name,
-                      -- then we need to use it in the package path.
-                      "../" <> (case Map.lookup pkgId pm of
-                                   Just (Just name, _) -> unPackageName name
-                                   Just (Nothing, _) -> unPackageId pkgId
-                                   Nothing -> error "IMPOSSIBLE : package map malformed"
-                               ) <> "/"
-            , let modNameStr = genModuleRef modRef
-            ]
-        defs = map (\(def, ser) -> def ++ ser) defSers
-    in
-    Just $ T.unlines $ intercalate [""] $ filter (not . null) $ header : imports : defs
+    let (defSers, refs) = unzip (map (genDataDef curPkgId mod tpls) serDefs)
+        imports = [ importDecl pm modRef rootPath
+                  | modRef@(pkgRef, _) <- modRefs refs
+                  , let rootPath = pkgRootPath modName pkgRef ]
+        defs = map biconcat defSers
+        modText = T.unlines $ intercalate [""] $ filter (not . null) $ modHeader : imports : defs
+        depends = [ T.unpack $ packageSetName <> "/" <> pkgRefStr pm pkgRef
+                  | (pkgRef, _) <- modRefs refs, isPkgRefNotSelf pkgRef ]
+   in Just (modText, depends)
   where
+    modName = moduleName mod
+    tpls = moduleTemplates mod
     serDefs = defDataTypes mod
+    modRefs refs = Set.toList ((PRSelf, modName) `Set.delete` Set.unions refs)
+    modHeader =
+      [ "// Generated from " <> T.intercalate "/" (unModuleName modName) <> ".daml"
+      , "/* eslint-disable @typescript-eslint/camelcase */"
+      , "/* eslint-disable @typescript-eslint/no-use-before-define */"
+      , "import * as jtv from '@mojotech/json-type-validation';"
+      , "import * as daml from '@daml/types';"
+      ]
+
+    -- Calculate an import declaration.
+    importDecl :: Map.Map PackageId (Maybe PackageName, Package) ->
+                       (PackageRef, ModuleName) -> T.Text -> T.Text
+    importDecl pm modRef@(pkgRef, modName) rootPath =
+      "import * as " <>  genModuleRef modRef <> " from '" <>
+      (if isPkgRefNotSelf pkgRef
+         then rootPath <> "/" <> pkgRefStr pm pkgRef <> "/lib/"
+         else rootPath <> "/" <> pkgRefStr pm pkgRef
+      ) <> T.intercalate "/" (unModuleName modName) <> "';"
+
+    -- Check if a package ref is a foreign import.
+    isPkgRefNotSelf :: PackageRef -> Bool
+    isPkgRefNotSelf = \case PRSelf -> False; PRImport _ -> True
+
+    -- Produce a package name for a package ref.
+    pkgRefStr :: Map.Map PackageId (Maybe PackageName, Package) ->
+                                              PackageRef -> T.Text
```

Comment:
```suggestion
    pkgRefStr :: Map.Map PackageId (Maybe PackageName, Package) -> PackageRef -> T.Text
```
The indentation here looks very odd.

Reply:
thought that would get flagged :) i have peculiar tastes. i'll normalize it as you suggest.